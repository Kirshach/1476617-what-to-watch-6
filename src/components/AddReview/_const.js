const Rating = `rating`;
const Comment = `comment`;

export const INITIAL_RATING = null;
export const INITAL_COMMENT = ``;

export const INITIAL_STATE = {
  [Rating]: INITIAL_RATING,
  [Comment]: INITAL_COMMENT,
};

export const FIELD_NAMES = {
  Rating,
  Comment,
};
