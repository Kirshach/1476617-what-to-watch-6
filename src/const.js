export const APIRoutes = {
  PROMO_FILM: `/films/promo`,
  FAVOURITE: `/favorite`,
  COMMENTS: `/comments`,
  LOGOUT: `/logout`,
  FILMS: `/films`,
  LOGIN: `/login`,

  getFilmRoute(id) {
    return [this.FILMS, id].join(`/`);
  },

  getCommentsRoute(id) {
    return [this.COMMENTS, id].join(`/`);
  },

  getFilmStatusRoute(id, isFavourite) {
    const status = isFavourite ? `0` : `1`;
    return [this.FAVOURITE, id, status].join(`/`);
  }
};

export const FilmSubroutes = {
  overview: ``,
  details: `details`,
  reviews: `reviews`,
  addReview: `review`,
};

export const AppRoutes = {
  MOVIE_PLAYER_BASE_ROUTE: `/player`,
  ADD_REVIEW: `/films/:id/review`,
  MOVIE_PLAYER: `/player/:id`,
  FILM_BASE_ROUTE: `/films`,
  FILM: `/films/:id/:tab?`,
  PAGE_NOT_FOUND: `/404`,
  MY_LIST: `/mylist`,
  LOGOUT: `/logout`,
  LOGIN: `/login`,
  MAIN: `/`,

  getFilmRoute(id, subroute) {
    return (subroute ? [this.FILM_BASE_ROUTE, id, subroute] : [this.FILM_BASE_ROUTE, id]).join(`/`);
  },

  getAddReviewRoute(id) {
    return this.getFilmRoute(id, FilmSubroutes.addReview);
  },

  getFilmReviewsRoute(id) {
    return this.getFilmRoute(id, FilmSubroutes.reviews);
  },

  getFilmDetailsRoute(id) {
    return this.getFilmRoute(id, FilmSubroutes.details);
  },

  getFilmOverviewRoute(id) {
    return this.getFilmRoute(id, FilmSubroutes.overview);
  },

  gerMoviePlayerRoute(id) {
    return [this.MOVIE_PLAYER_BASE_ROUTE, id].join(`/`);
  },
};

export const ALL_GENRES = `All genres`;
