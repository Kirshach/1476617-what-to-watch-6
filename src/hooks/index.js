export {useFavouriteFilms} from './use-favourite-films';
export {useNavigation} from './use-navigation';
export {usePromo} from './use-promo';
export {useFilms} from './use-films';
export {useForm} from './use-form';
export {useAuth} from './use-auth';
