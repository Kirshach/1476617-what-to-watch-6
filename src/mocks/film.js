export const macbeth = {
  name: `Macbeth`,
  posterImage: `https://assets.htmlacademy.ru/intensives/javascript-3/film/poster/Macbeth.jpg`,
  previewImage: `https://assets.htmlacademy.ru/intensives/javascript-3/film/preview/macbeth.jpg`,
  backgroundImage: `https://assets.htmlacademy.ru/intensives/javascript-3/film/background/Macbeth.jpg`,
  backgroundColor: `#F1E9CE`,
  description: `Macbeth, the Thane of Glamis, receives a prophecy from a trio of witches that one day he will become King of Scotland. Consumed by ambition and spurred to action by his wife, Macbeth murders his king and takes the throne for himself.`,
  rating: 3.3,
  scoresCount: 48798,
  director: `Justin Kurzel`,
  starring: [
    `Michael Fassbender`,
    `Marion Cotillard`,
    `Jack Madigan`
  ],
  runTime: 113,
  genre: `Drama`,
  released: 2015,
  id: 1,
  isFavorite: true,
  videoLink: `http://peach.themazzone.com/durian/movies/sintel-1024-surround.mp4`,
  previewVideoLink: `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`
};

export const bronson = {
  name: `Bronson`,
  posterImage: `https://assets.htmlacademy.ru/intensives/javascript-3/film/poster/bronson.jpg`,
  previewImage: `https://assets.htmlacademy.ru/intensives/javascript-3/film/preview/bronson.jpg`,
  backgroundImage: `https://assets.htmlacademy.ru/intensives/javascript-3/film/background/bronson.jpg`,
  backgroundColor: `#73B39A`,
  description: `A young man who was sentenced to seven years in prison for robbing a post office ends up spending three decades in solitary confinement. During this time, his own personality is supplanted by his alter-ego, Charles Bronson.`,
  rating: 3.6,
  scoresCount: 109661,
  director: `Nicolas Winding Refn`,
  starring: [
    `Tom Hardy`,
    `Kelly Adams`,
    `Luing Andrews`
  ],
  runTime: 92,
  genre: `Action`,
  released: 2008,
  id: 9,
  isFavorite: false,
  videoLink: `http://peach.themazzone.com/durian/movies/sintel-1024-surround.mp4`,
  previewVideoLink: `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`
};

export const revenant = {
  name: `The Revenant`,
  posterImage: `https://assets.htmlacademy.ru/intensives/javascript-3/film/poster/Revenant.jpg`,
  previewImage: `https://assets.htmlacademy.ru/intensives/javascript-3/film/preview/revenant.jpg`,
  backgroundImage: `https://assets.htmlacademy.ru/intensives/javascript-3/film/background/Revenant.jpg`,
  backgroundColor: `#92918B`,
  description: `A frontiersman on a fur trading expedition in the 1820s fights for survival after being mauled by a bear and left for dead by members of his own hunting team.`,
  rating: 4,
  scoresCount: 618498,
  director: `Alejandro G. Iñárritu`,
  starring: [
    `Leonardo DiCaprio`,
    `Tom Hardy`,
    `Will Poulter`
  ],
  runTime: 156,
  genre: `Action`,
  released: 2015,
  id: 8,
  isFavorite: false,
  videoLink: `http://peach.themazzone.com/durian/movies/sintel-1024-surround.mp4`,
  previewVideoLink: `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`
};
