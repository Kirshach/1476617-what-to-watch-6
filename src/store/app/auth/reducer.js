import {ActionType} from './actions';

const HAS_CHECKED_AUTH = `hasCheckedAuth`;
const IS_AUTHORIZED = `isAuthorized`;
const USER_DATA = `userData`;

const initialState = {
  [HAS_CHECKED_AUTH]: false,
  [IS_AUTHORIZED]: false,
  [USER_DATA]: {
    avatarUrl: null,
    email: null,
    name: null,
    id: null,
  },
};

export const authReducer = (state = initialState, action) => {
  switch (action.type) {
    case ActionType.SET_IS_AUTHORIZED:
      return {
        ...state,
        [IS_AUTHORIZED]: action.payload,
      };
    case ActionType.SET_USER_DATA:
      return {
        ...state,
        [USER_DATA]: action.payload,
      };
    case ActionType.SET_HAS_CHECKED_AUTH:
      return {
        ...state,
        [HAS_CHECKED_AUTH]: action.payload,
      };
    default:
      return state;
  }
};

export const AuthNS = {
  HAS_CHECKED_AUTH,
  IS_AUTHORIZED,
  USER_DATA,
};
