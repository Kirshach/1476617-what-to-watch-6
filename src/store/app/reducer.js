import {ActionType} from './actions';
import {Genres} from '../../const';

export const FILMS_BATCH = 8;

const initialState = {
  filmsShowingCount: FILMS_BATCH,
  genre: Genres.allGenres,
  isLoadingFilms: true,
};

export const appReducer = (state = initialState, action) => {
  switch (action.type) {
    case ActionType.SET_GENRE:
      return {
        ...state,
        genre: action.payload
      };
    case ActionType.SET_IS_LOADING_FILMS:
      return {
        ...state,
        isLoadingFilms: action.payload
      };
    case ActionType.SHOW_MORE_FILMS:
      return {
        ...state,
        filmsShowingCount: state.filmsShowingCount + (action.payload || FILMS_BATCH),
      };
    case ActionType.SHOW_DEFAULT_FILMS_BATCH:
      return {
        ...state,
        filmsShowingCount: state.filmsShowingCount + (action.payload || FILMS_BATCH),
      };
    default:
      return state;
  }
};
