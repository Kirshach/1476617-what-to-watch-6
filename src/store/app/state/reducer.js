import {Genres} from '../../../const';
import {ActionType} from './actions';

export const FILMS_BATCH = 8;

const REVIEWS_HAVE_LOADED = `reviewsHaveLoaded`;
const FILMS_SHOWING_COUNT = `filmsShowingCount`;
const FILMS_HAVE_LOADED = `filmsHaveLoaded`;
const PROMO_HAS_LOADED = `promoHasLoaded`;
const FILM_HAS_LOADED = `filmHasLoaded`;
const GENRE = `genre`;

const initialState = {
  [REVIEWS_HAVE_LOADED]: false,
  [FILMS_HAVE_LOADED]: false,
  [PROMO_HAS_LOADED]: false,
  [FILM_HAS_LOADED]: false,

  [FILMS_SHOWING_COUNT]: FILMS_BATCH,
  [GENRE]: Genres.allGenres,
};

export const stateReducer = (state = initialState, action) => {
  switch (action.type) {
    case ActionType.SET_FILM_HAS_LOADED:
      return {
        ...state,
        [FILM_HAS_LOADED]: action.payload,
      };
    case ActionType.SET_FILMS_HAVE_LOADED:
      return {
        ...state,
        [FILMS_HAVE_LOADED]: action.payload,
      };
    case ActionType.SET_PROMO_HAS_LOADED:
      return {
        ...state,
        [PROMO_HAS_LOADED]: action.payload,
      };
    case ActionType.SET_GENRE:
      return {
        ...state,
        [GENRE]: action.payload,
      };
    case ActionType.SET_REVIEWS_HAVE_LOADED:
      return {
        ...state,
        [REVIEWS_HAVE_LOADED]: action.payload,
      };
    case ActionType.SHOW_DEFAULT_FILMS_AMOUNT:
      return {
        ...state,
        [FILMS_SHOWING_COUNT]: FILMS_BATCH,
      };
    case ActionType.SHOW_MORE_FILMS:
      return {
        ...state,
        [FILMS_SHOWING_COUNT]: state[FILMS_SHOWING_COUNT] + (action.payload || FILMS_BATCH),
      };
    default:
      return state;
  }
};

export const StateNS = {
  REVIEWS_HAVE_LOADED,
  FILMS_SHOWING_COUNT,
  FILMS_HAVE_LOADED,
  PROMO_HAS_LOADED,
  FILM_HAS_LOADED,
  GENRE,
};
