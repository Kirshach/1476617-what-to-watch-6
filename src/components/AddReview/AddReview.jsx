import React, {Fragment} from 'react';
import {Link, Redirect} from 'react-router-dom';

import {withFilms, withFilmsPropTypes} from '../../hocs/withFilms';
import {useQueryFilmById} from '../../hooks/useQueryFilmById';
import {useForm} from '../../hooks/useForm.js';

import LoadingPlaceholder from '../LoadingPlaceholder/LoadingPlaceholder';
import PageHeader from '../PageHeader/PageHeader';

const INITIAL_RATING = 8;
const INITIAL_STATE = {
  rating: INITIAL_RATING,
  comment: ``
};

export const AddReview = ({films, filmsHaveLoaded}) => {
  const film = useQueryFilmById(films);

  const {values, handlers} = useForm(INITIAL_STATE);
  const {comment} = values;
  const onSubmit = (evt) => {
    evt.preventDefault();
    // eslint-disable-next-line no-alert
    alert(JSON.stringify(values));
  };

  if (filmsHaveLoaded && !film.id) {
    return <Redirect to="/404" />;
  }

  return filmsHaveLoaded
    ? (
      <section className="movie-card movie-card--full">
        <div className="movie-card__header">
          <div className="movie-card__bg">
            <img src="img/bg-the-grand-budapest-hotel.jpg" alt="The Grand Budapest Hotel" />
          </div>

          <h1 className="visually-hidden">WTW</h1>

          <PageHeader>
            <nav className="breadcrumbs">
              <ul className="breadcrumbs__list">
                <li className="breadcrumbs__item">
                  <Link to={`films/${film.id}`} className="breadcrumbs__link">{film.name}</Link>
                </li>
                <li className="breadcrumbs__item">
                  <a className="breadcrumbs__link">Add review</a>
                </li>
              </ul>
            </nav>
          </PageHeader>

          <div className="movie-card__poster movie-card__poster--small">
            <img src={film.posterImage} alt={`${film.name} poster`} width="218" height="327" />
          </div>
        </div>

        <div className="add-review">
          <form action="#" className="add-review__form" onSubmit={onSubmit}>
            <div className="rating">
              <div className="rating__stars">
                {new Array(10).fill(null).map((_, index) => {
                  const starsCount = index + 1;
                  return (
                    <Fragment key={starsCount}>
                      <input className="rating__input" id={`star-${starsCount}`} type="radio" name="rating" value={starsCount} onChange={handlers.rating}/>
                      <label className="rating__label" htmlFor={`star-${starsCount}`}>Rating {starsCount}</label>
                    </Fragment>
                  );
                })}
              </div>
            </div>

            <div className="add-review__text">
              <textarea
                className="add-review__textarea"
                name="review-text"
                id="review-text"
                placeholder="Review text"
                onChange={handlers.comment}
                value={comment}
              />
              <div className="add-review__submit">
                <button className="add-review__btn" type="submit">Post</button>
              </div>

            </div>
          </form>
        </div>

      </section>
    ) : (
      <LoadingPlaceholder />
    );
};

AddReview.propTypes = withFilmsPropTypes;

const AddReviewWithFilms = withFilms(AddReview);

export default AddReviewWithFilms;
