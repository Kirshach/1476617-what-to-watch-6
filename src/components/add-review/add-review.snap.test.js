import React from 'react';
import {useDispatch, useSelector} from 'react-redux';
import {MemoryRouter, Route} from 'react-router-dom';
import {render} from '@testing-library/react';

import {INITIAL_STATE, DomainNS} from '../../store/domain/reducer';
import {store} from '../../mocks/store/store';
import {useFilm} from '../../hooks/use-film';
import {macbeth} from '../../mocks/film';
import {AppRoutes} from '../../const';

import AddReview from './add-review';

jest.mock(`react-redux`);
jest.mock(`../../hooks/use-film`);

describe(`<AddReview />`, () => {
  const {id} = macbeth;
  useDispatch.mockReturnValue(() => Promise.resolve());
  useSelector.mockImplementation((selector) => selector(store));

  test(`renders correctly for a film that has loaded`, () => {
    useFilm.mockReturnValueOnce([macbeth, true]);

    const {container} = render(
        <MemoryRouter initialEntries={[AppRoutes.getAddReviewRoute(id)]}>
          <Route path={AppRoutes.ADD_REVIEW}>
            <AddReview />
          </Route>
        </MemoryRouter>
    );
    expect(container).toMatchSnapshot();
  });

  test(`renders correctly for a film that has not loaded`, () => {
    const initialFilm = INITIAL_STATE[DomainNS.FILM];
    useFilm.mockReturnValueOnce([initialFilm, false]);

    const {container} = render(
        <MemoryRouter initialEntries={[AppRoutes.getAddReviewRoute(id)]}>
          <AddReview />
        </MemoryRouter>
    );
    expect(container).toMatchSnapshot();
  });
});
