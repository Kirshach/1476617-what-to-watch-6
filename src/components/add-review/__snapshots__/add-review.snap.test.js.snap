// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AddReview /> renders correctly for a film that has loaded 1`] = `
<div>
  <section
    class="movie-card movie-card--full"
  >
    <div
      class="movie-card__header"
    >
      <div
        class="movie-card__bg"
      >
        <img
          alt="The Grand Budapest Hotel"
          src="img/bg-the-grand-budapest-hotel.jpg"
        />
      </div>
      <h1
        class="visually-hidden"
      >
        WTW
      </h1>
      <header
        class="page-header "
      >
        <div
          class="logo"
        >
          <a
            class="logo__link"
            href="/"
          >
            <span
              class="logo__letter logo__letter--1"
            >
              W
            </span>
            <span
              class="logo__letter logo__letter--2"
            >
              T
            </span>
            <span
              class="logo__letter logo__letter--3"
            >
              W
            </span>
          </a>
        </div>
        <nav
          class="breadcrumbs"
        >
          <ul
            class="breadcrumbs__list"
          >
            <li
              class="breadcrumbs__item"
            >
              <a
                class="breadcrumbs__link"
                href="/films/1"
              >
                Macbeth
              </a>
            </li>
            <li
              class="breadcrumbs__item"
            >
              <a
                class="breadcrumbs__link"
              >
                Add review
              </a>
            </li>
          </ul>
        </nav>
        <div
          class="user-block"
        >
          <a
            class="user-block__link"
            href="/mylist"
          >
            <div
              class="user-block__avatar"
            >
              <img
                alt="User avatar"
                height="63"
                src="https://assets.htmlacademy.ru/intensives/javascript-3/avatar/6.jpg"
                width="63"
              />
            </div>
          </a>
        </div>
      </header>
      <div
        class="movie-card__poster movie-card__poster--small"
      >
        <img
          alt="Macbeth poster"
          height="327"
          src="https://assets.htmlacademy.ru/intensives/javascript-3/film/poster/Macbeth.jpg"
          width="218"
        />
      </div>
    </div>
    <div
      class="add-review"
    >
      <form
        action="#"
        class="add-review__form"
      >
        <div
          class="rating"
        >
          <div
            class="rating__stars"
          >
            <input
              class="rating__input"
              id="star-1"
              name="rating"
              type="radio"
              value="1"
            />
            <label
              class="rating__label"
              for="star-1"
            >
              Rating 
              1
            </label>
            <input
              class="rating__input"
              id="star-2"
              name="rating"
              type="radio"
              value="2"
            />
            <label
              class="rating__label"
              for="star-2"
            >
              Rating 
              2
            </label>
            <input
              class="rating__input"
              id="star-3"
              name="rating"
              type="radio"
              value="3"
            />
            <label
              class="rating__label"
              for="star-3"
            >
              Rating 
              3
            </label>
            <input
              class="rating__input"
              id="star-4"
              name="rating"
              type="radio"
              value="4"
            />
            <label
              class="rating__label"
              for="star-4"
            >
              Rating 
              4
            </label>
            <input
              class="rating__input"
              id="star-5"
              name="rating"
              type="radio"
              value="5"
            />
            <label
              class="rating__label"
              for="star-5"
            >
              Rating 
              5
            </label>
            <input
              class="rating__input"
              id="star-6"
              name="rating"
              type="radio"
              value="6"
            />
            <label
              class="rating__label"
              for="star-6"
            >
              Rating 
              6
            </label>
            <input
              class="rating__input"
              id="star-7"
              name="rating"
              type="radio"
              value="7"
            />
            <label
              class="rating__label"
              for="star-7"
            >
              Rating 
              7
            </label>
            <input
              class="rating__input"
              id="star-8"
              name="rating"
              type="radio"
              value="8"
            />
            <label
              class="rating__label"
              for="star-8"
            >
              Rating 
              8
            </label>
            <input
              class="rating__input"
              id="star-9"
              name="rating"
              type="radio"
              value="9"
            />
            <label
              class="rating__label"
              for="star-9"
            >
              Rating 
              9
            </label>
            <input
              class="rating__input"
              id="star-10"
              name="rating"
              type="radio"
              value="10"
            />
            <label
              class="rating__label"
              for="star-10"
            >
              Rating 
              10
            </label>
          </div>
        </div>
        <div
          class="add-review__text"
        >
          <textarea
            class="add-review__textarea"
            id="review-text"
            name="review-text"
            placeholder="Review text"
          />
          <div
            class="add-review__submit"
          >
            <button
              class="add-review__btn"
              type="submit"
            >
              Post
            </button>
          </div>
        </div>
      </form>
    </div>
  </section>
</div>
`;

exports[`<AddReview /> renders correctly for a film that has not loaded 1`] = `
<div>
  <div
    class="loading-placeholder"
  >
    Loading
    .
  </div>
</div>
`;
